### PR TITLE
refactor(api): move engine execution logic closer to req/res payloads

### PR DESCRIPTION
**Pull request by mcous**
## Overview

This PR is a reorganization that was important to get in before #6596, #6598, and #6599 (otherwise it would be even bigger).

This PR reorganizes where the execution logic of a given command lives in order to scale more gracefully as we continue to add functionality. "Execution logic", in this case, means the side-effects of a given command (e.g. calls to the hardware controller) and _not_ how the command affects the ProtocolEngine's state.

## Changelog

Before this PR, to execute a command's side effects, there was a single monolithic `CommandExecutor` that acted as a router. All commands would come into the `CommandExecutor`, and it would identify the command and send it off to a specific handler (e.g. `PipettingHandler`) for processing. This had the following downsides:

1. The `CommandExecutor` required knowledge of all available commands
2. The `CommandExecutor` and its handlers used the command request models as input, coupling those interfaces together
3. As commands were added to the engine, the `CommandExecutor` class grew linearly, and it was already unwieldy with 4 total commands
4. Because the command executor operated on command models directly, it started to dangerously mix state changes with side effect execution

The system had the following upsides:

5. Command request and response models were not tied to the implementation details past the level of including enough data in the model to successfully execute
6. At the handler level, side effect logic was well separated from state logic in a manner that was easy to reason with and test

In order to eliminate / reduce the downsides above while preserving the upsides, the new system:

- Breaks down each command into three parts:
    - A request data model (almost unchanged)
    - A result data model (unchanged)
    - An implementation class
- The implementation class:
    - Is a subclass of `CommandImplementation`
    - Has a `create_command` method that creates `PendingCommand` model from the request
        - Mitigates downside 4
    - Has an `execute` method to trigger the request's side effects
        - Mitigates downside 3
- The `execute` method of `CommandImplementation` is passed a `CommandHandlers` instance
    - `CommandHandlers` is a `CommandExecutor`, but all routing logic has been removed, so that it's simply the collection of all the different command handling logic
        - Preserves upside 6
    - The existing handlers have been reworked to operate on function parameters rather than request models
        - Mitigates downside 2
- All command requests now have a `get_implementation()` method that returns that command's specific command implementation
    - By allowing the request to retrieve its own implementation, we remove the need for any sort of routing layer
        - Mitigates downsides 1 and 3
    - By not adding the implementation logic itself to the request, though, we preserve upside 5

### Commands

As a recap, there are four commands the engine currently support. They are now located at:

- Load labware: `protocol_engine/commands/load_labware.py`
- Load pipette: `protocol_engine/commands/load_pipette.py`
- Move to well: `protocol_engine/commands/move_to_well.py`
- Pick up tip: `protocol_engine/commands/pick_up_tip.py`

Each of these files contain a request model, a result model, and a command implementation.

### Command execution

A couple things to note here as you're reading through it:

- The diff for the `command_handlers.py` file looks big in git, but really its a rename + edit of the removed `command_excutor.py` file
- The bulk of the rest of the changes here are around the handler / request model decoupling

### Tests

- We were missing tests for the command state models (`PendingCommand`, `RunningCommand`, etc.)
- Tests that were in the `CommandExecutor` tests are now spread out between the protocol engine and the command implementation tests

### Smoke testing

Smoke test Jupyter Notebook protocol: https://github.com/Opentrons/opentrons/pull/7026#issuecomment-729920713

I'm able to use this protocol to load a pipette, load two tip racks, move to a tip rack, and pick up a tip from the other tip rack.

## Risk assessment

Very low, especially once the smoke testing plan is posted. The logic is not hooked to anything, but it has test coverage in the 90s.

### CODE REVIEW

File: api/src/opentrons/protocol_engine/commands/__init__.py [hunk 1]
```diff
@@ -138,12 +100,4 @@
     "MoveToWellResult",
     "PickUpTipRequest",
     "PickUpTipResult",
-    "DropTipRequest",
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/commands/__init__.py:141]: Removed these until we're actually using them, because otherwise this PR would've gotten even bigger

File: api/src/opentrons/protocol_engine/commands/command.py [hunk 2]
```diff
@@ -1,12 +1,19 @@
...
-from typing import Generic, TypeVar, Union
+from typing import TYPE_CHECKING, Generic, TypeVar
 
 from ..errors import ProtocolEngineError
 
+# convenience type alias to work around type-only circular dependency
+if TYPE_CHECKING:
+    from ..execution import CommandHandlers
+else:
+    CommandHandlers = None
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/commands/command.py:15]: This is here so individual command implementation modules can import `CommandHandlers` without the `TYPE_CHECKING` guard

**Review by sfoster1 — COMMENTED** [review]
So is the workflow for creating a new command now
- Make a new module for the command in `protocol_engine.commands.my_command`
- Define its request model, its response model, and its implementation in `protocol_engine.command.my_command`
- Define an action handler and hook it up in `protocol_engine.execution`
- Add its command type to the pending, running, completed, and failed unions in `protocol_engine.command_models.__init__`
- Add state handler entries to the handle functions in the right sub-entry of `protocol_engine.state.motion`

is that pretty much right?

If so, can we have an equivalent to `get_implementation` for the state handlers so there's less boilerplate and fewer files need to be touched? I know you want to keep changes to the lifecycle state and actually emitting hardware commands separate, but this makes it really hard to have a specific place to look to understand the behavior of a part of the system, and makes essentially duplicated logic around matching commands in a couple places. 

There's also an interesting set of stuff we could do with metaclass registration that could similarly ease the burden. Specifically, we should look into using [subclass registration](https://www.python.org/dev/peps/pep-0487/#subclass-registration) when defining models to try and pull new models automatically. I'm not sure how introducing class hierarchies will affect pydantic model differentiation, but if we
- used subclass registration in the model
- had a similar approach to finding state handlers from the model class like we do for command execution

then there's a good chance we could narrow down the files you have to touch to add a command (and thus the files you need to look at to understand or find bugs in a command) to just the new command submodel in `protocol_engine.commands`.

I think there's a way we could change the handlers to not be so separate as well along similar lines. Commands own their handlers; handlers are either not registered and are free functions that take the arguments for the state stuff they need, or are registered similarly as lookups from command models.

Basically, my only problem with this is that what each command does is spread around a bunch of different files, and I think we can leverage a combination of subclass registration for request models and lookups on the request model classes for dependent handlers to fix it.

**Comment by mcous** [comment]
Added context:

> So is the workflow for creating a new command now
>
> - Make a new module for the command in `protocol_engine.commands.my_command`

👍 

> - Define its request model, its response model, and its implementation in `protocol_engine.command.my_command`

👍 

> - Define an action handler and hook it up in `protocol_engine.execution`

👍, with caveats:

There's a pretty good chance this one-to-one relationship between command implementation and handler logic won't hold up, and more logic will shift to the command implementation. You could imagine a change where:

- Currently, `PickUpTipImplementation` calls `handlers.pipetting.pick_up_tip`, which internally calls `movement.move_to_well`
- Later, we refactor so the implementation chains `pipetting.pick_up_tip` and `movement.move_to_well` itself

We're taking a wait-and-see approach, with an eye towards trying to make the handlers more composable and focused. Command validation is probably going to factor into this somehow (and potentially involve some sort of chaining/builder pattern for command handlers).

> - Add its command type to the pending, running, completed, and failed unions in `protocol_engine.command_models.__init__`

Also 👍, with caveats:

These classes and unions are unwieldy. I really like separate classes to represent state changes, but I also think it's a pattern that's more helpful in strongly typed languages than it is here in Python. Here, I worry the benefits aren't outweighing the awkwardness. We could, for example, have a single command class with a status string and various optional fields, where the downside would be that we wouldn't know at type linting time that `command.status == completed` means `command.result` is definitely defined.

> - Add state handler entries to the handle functions in the right sub-entry of `protocol_engine.state.motion`

I would rephrase this to "add state handler entries to **any substores that need to react to that particular command**, if applicable."

The key difference here is that there is always a one-to-one relationship between a given command and a given execution. There is a one-to-zero-or-more relationship between a successful command result and a substate change.

> If so, can we have an equivalent to get_implementation for the state handlers so there's less boilerplate and fewer files need to be touched? I know you want to keep changes to the lifecycle state and actually emitting hardware commands separate, but this makes it really hard to have a specific place to look to understand the behavior of a part of the system, and makes essentially duplicated logic around matching commands in a couple places.

I'm not comfortable with this change, and I'm not sure how it would look without spreading state mutation across the engine codebase, which is the exact thing this architecture is designed to avoid.

I also want to push back gently on "this makes it really hard to have a specific place to look to understand the behavior of a part of the system". I think the engine is optimized to answer the question differently than our code currently does.

To take an example question: "how do the engine decide its current location?"

1. `ProtcolContext`
    1. Find `ctx._location_cache`
    2. Find the places that modify it in instrument context
    3. Examine how those places mutate the value
2. Protocol engine with `impl.update_state` or similar strawman
    1. Find `motion` substore
    2. Find the commands that touch the motion substate in their `update_state` methods
    3. Examine how those places mutate the value
3. Protocol engine as written
    1. Find the `motion` substore
    2. You're done, how and why the state changes is explicitly declared in the store

> Basically, my only problem with this is that what each command does is spread around a bunch of different files

For me, this is a feature and not a bug 😄. Another way to look at this might be: the logic for the creation and execution of a single command flows through several files, but the flow is well defined and each of those files is small, focused, explicit, and easily testable at their boundaries.

- A command file defines a request model, a result model, and a way to compose execution handlers to "map" the request into a result
- A handler defines individual, focused actions in a given side-effect subdomain that know how to read required inputs from state and can be composed
- A store defines a reactive data model for a state subdomain and explicit guarantees about how and why that data may change

~Boilerplate on its own is not an enemy of reliability; tight coupling usually is.~ < Upon further reflection, this is a nonsense statement.

Another thing to note is that by trading in a little bit of boilerplate and file jumping into the "add a new command to the engine" flow, we gain the ability to add other types of complex features in a more focused manner. For example, adding basic liquid tracking functionality could be done by adding a single `liquid_tracking` substore that reacts to already existing liquid handling commands.

> I think we can leverage a combination of subclass registration for request models and lookups on the request model classes for dependent handlers to fix it.

Subclass registration looks interesting (depending on how much "magic" is involved), and I agree wholeheartedly that a generalized plugin system is the way to take the engine and its functionality.

File: api/src/opentrons/protocol_engine/protocol_engine.py [hunk 3]
```diff
@@ -52,23 +53,35 @@ def __init__(
...
         """
         self.state_store = state_store
-        self.executor = executor
+        self._handlers = handlers
 
     async def execute_command(
         self,
         request: CommandRequestType,
         command_id: str,
     ) -> Union[CompletedCommandType, FailedCommandType]:
```
Comment (amitlissack) [inline api/src/opentrons/protocol_engine/protocol_engine.py:62]: We talked about it in our review but I wanted to write a comment here for posterity. 

`CommandRequestType` is a `Union` of all request types. In reality `request`'s type can be "anything that has a `get_implementation` method." We could have an `ABC` defining `get_implementation` which request models implement. It might make things more clear. I'm on the fence.

**Review by amitlissack — APPROVED** [review]
I am very fond of this PR. I'll take more small source files over fewer large source files.

File: api/src/opentrons/protocol_engine/protocol_engine.py (hunk 3 above)
Comment (mcous) [reply api/src/opentrons/protocol_engine/protocol_engine.py:62]: Yeah, agreed that this needs some work. I'm not sure I have any good ideas for this PR specifically, which doesn't touch this signature. A few more points:

- The `CommandRequestType` also declares "this is the finite set of commands that the ProtocolEngine knows about"
     - This will not scale if engine plugins and custom commands are ever a thing, which I think is a mark in favor of dropping to an abstract declaration
- It would be nice to have some way of having the type system somehow know that if you give it a specific request, you're going to get back a command resource of that specific request and result shape
    - It would be nice if we could avoid spamming the protocol engine methods with a dozen `@overload`s to handle every single known command type

**Comment by mcous** [comment]
Testing protocol for Jupyter Notebook:

```py
import asyncio
import opentrons.execute
from uuid import uuid4
from typing import Optional

from opentrons.types import DeckSlotName, MountType
from opentrons.protocol_api import ProtocolContext
from opentrons.protocol_engine import ProtocolEngine, commands
from opentrons.protocol_engine.types import DeckSlotLocation

opentrons.execute._clear_cached_hardware_controller()
protocol = opentrons.execute.get_protocol_api('2.7')
protocol.home()

loop = asyncio.get_event_loop()
loop.create_task(run_protocol(protocol))

async def run_protocol(ctx: ProtocolContext) -> None:
    # HACK(mc, 2020-11-06): don't ever do this in a real protocol
    hw_api = ctx._hw_manager.hardware._obj_to_adapt
    engine = ProtocolEngine.create(hardware=hw_api)

    async def execute_command(
        request: commands.CommandRequestType
    ) -> Optional[commands.CommandResultType]:
        command_id = str(uuid4())
        command = await engine.execute_command(request, command_id=command_id)
        print(f"{type(command.request).__name__} - {command_id}")
        print(f"{command.request}")

        if isinstance(command, commands.FailedCommand):
            print(f"error: {command.error}")
            return None

        print(f"result: {command.result}")
        return command.result

    result = await execute_command(
        commands.LoadLabwareRequest(
            location=DeckSlotLocation(DeckSlotName.SLOT_5),
            loadName="opentrons_96_tiprack_300ul",
            namespace="opentrons",
            version=1
        ),
    )

    rack_5 = result.labwareId  # type: ignore[union-attr]

    result = await execute_command(
        commands.LoadLabwareRequest(
            location=DeckSlotLocation(DeckSlotName.SLOT_8),
            loadName="opentrons_96_tiprack_300ul",
            namespace="opentrons",
            version=1
        ),
    )

    rack_8 = result.labwareId  # type: ignore[union-attr]

    result = await execute_command(
        commands.LoadPipetteRequest(
            pipetteName="p300_single_gen2",
            mount=MountType.RIGHT,
        ),
    )

    p300 = result.pipetteId  # type: ignore[union-attr]

    await execute_command(
        commands.MoveToWellRequest(pipetteId=p300, labwareId=rack_5, wellName="A1")
    )

    await execute_command(
        commands.PickUpTipRequest(pipetteId=p300, labwareId=rack_8, wellName="H12")
    )
```
